<a href="https://colab.research.google.com/github/panchaldhruv27223/DA_EDA/blob/main/EDA_Lab_Assignment_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Name** : Dhruv Panchal
### **SID** : *202411042*


In [2]:
from google.colab import drive
drive.mount("/content/drive")

Mounted at /content/drive


# Dataset Link: https://www.kaggle.com/datasets/mrmars1010/imdb-top250-india


#### download dataset from the given link and upload to the drive.

In [3]:
from google.colab import files
uploaded = files.upload()

Saving IMdB_India_Top250.csv to IMdB_India_Top250.csv


### **Q 1. Data Inspection**

  > a. Load the dataset into a pandas DataFrame and display the first 10 rows.

  > b. Use the info() function to get an overview of the dataset, including data types and non-null counts.

  > c. Display the summary statistics for numerical columns using describe().


In [4]:
import pandas as pd
import numpy as np

In [5]:
df = pd.read_csv("/content/IMdB_India_Top250.csv")
df.head(10)

Movie name  Year of release        Watch  hour   Rating Ratedby  \
0          12th Fail             2023  2 hours 27 minutes     8.9    126K   
1           Gol Maal             1979            2 hours      8.5     20K   
2           Maharaja             2024  2 hours 30 minutes     8.6     37K   
3            Nayakan             1987  2 hours 25 minutes     8.7     25K   
4   The World of Apu             1959   1 hour 45 minutes     8.4     17K   
5         Anbe Sivam             2003  2 hours 40 minutes     8.6     26K   
6  Pariyerum Perumal             2018  2 hours 34 minutes     8.7     19K   
7           3 Idiots             2009  2 hours 50 minutes     8.4    441K   
8              #Home             2021  2 hours 38 minutes     8.8     16K   
9       Black Friday             2004  2 hours 23 minutes     8.4     22K   

           Film Industry                   Genre              Director  \
0      Bollywood (Hindi)        Drama, Biography    Vidhu Vinod Chopra   
1      Bollywood (Hindi)                  Comedy  Hrishikesh Mukherjee   
2      Kollywood (Tamil)            Crime, Drama   Nithilan Saminathan   
3      Kollywood (Tamil)            Crime, Drama           Mani Ratnam   
4         Bengali Cinema                   Drama          Satyajit Ray   
5      Kollywood (Tamil)           Drama, Comedy             Sundar C.   
6      Kollywood (Tamil)           Drama, Social         Mari Selvaraj   
7      Bollywood (Hindi)           Comedy, Drama       Rajkumar Hirani   
8  Mollywood (Malayalam)           Drama, Family          Rojin Thomas   
9      Bollywood (Hindi)  Crime, Drama, Thriller        Anurag Kashyap   

  Box office collection  User reviews                        Awards  \
0           $138,288.00           945      23 wins & 32 nominations   
1                   NIL            48         3 wins & 1 nomination   
2           $975,543.00           370                 2 nominations   
3           $120,481.93           237         7 wins & 1 nomination   
4           $134,241.00            62  4 wins & 2 nominations total   
5                   NIL           115        2 wins & 3 nominations   
6                   NIL           169       11 wins & 5 nominations   
7        $60,262,836.00          1000      64 wins & 30 nominations   
8                   NIL           541        3 wins & 2 nominations   
9         $1,610,897.00            88         1 win & 7 nominations   

                                         Description  \
0  The real-life story of IPS Officer Manoj Kumar...   
1  A man's simple lie to secure his job escalates...   
2  A barber seeks vengeance after his home is bur...   
3  A common man's struggles against a corrupt pol...   
4  This final installment in Satyajit Ray's Apu T...   
5  Two men, one young and arrogant, the other dam...   
6  A law student from a lower caste begins a frie...   
7  Two friends are searching for their long lost ...   
8  Oliver Twist (Indrans) wants to be tech-savvy ...   
9  A film about the investigations following the ...   

                  Streaming platform  
0                            SonyLIV  
1  Amazon Prime Video, YouTube, Zee5  
2                            Netflix  
3        Amazon Prime Video, YouTube  
4        Amazon Prime Video, Hoichoi  
5                    Disney+ Hotstar  
6                 Amazon Prime Video  
7        Netflix, Amazon Prime Video  
8                 Amazon Prime Video  
9                            YouTube

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Movie name             250 non-null    object 
 1   Year of release        250 non-null    int64  
 2   Watch  hour            250 non-null    object 
 3   Rating                 250 non-null    float64
 4   Ratedby                250 non-null    object 
 5   Film Industry          250 non-null    object 
 6   Genre                  250 non-null    object 
 7   Director               250 non-null    object 
 8   Box office collection  250 non-null    object 
 9   User reviews           250 non-null    int64  
 10  Awards                 250 non-null    object 
 11  Description            250 non-null    object 
 12  Streaming platform     250 non-null    object 
dtypes: float64(1), int64(2), object(10)
memory usage: 25.5+ KB


In [7]:
len(df)

250

In [8]:
df.shape

(250, 13)

### dataset contain the 250 row(examples) and total 13 columns. and we are able to see that columns are not contaning the missing values.

## summary statistics

In [11]:
df.describe(include="number")

Year of release      Rating  User reviews
count       250.000000  250.000000     250.00000
mean       2008.388000    8.190400     222.18400
std          14.708597    0.252513     388.15949
min        1955.000000    7.700000      10.00000
25%        2003.000000    8.000000      39.00000
50%        2013.000000    8.150000      94.50000
75%        2018.000000    8.300000     231.50000
max        2024.000000    9.100000    3300.00000

In [12]:
df.dtypes

Movie name                object
Year of release            int64
Watch  hour               object
Rating                   float64
Ratedby                   object
Film Industry             object
Genre                     object
Director                  object
Box office collection     object
User reviews               int64
Awards                    object
Description               object
Streaming platform        object
dtype: object

### **Q 2. Data Cleaning**
  > a. Identify and handle any missing values in the dataset. Drop rows or fill missing values as appropriate.

  > b. Remove any duplicate rows from the dataset. Ensure that the columns related to rankings, years, or ratings are in the correct numerical format.

In [13]:
df.isnull().sum()

Movie name               0
Year of release          0
Watch  hour              0
Rating                   0
Ratedby                  0
Film Industry            0
Genre                    0
Director                 0
Box office collection    0
User reviews             0
Awards                   0
Description              0
Streaming platform       0
dtype: int64

In [16]:
df.dtypes

Movie name                object
Year of release            int64
Watch  hour               object
Rating                   float64
Ratedby                   object
Film Industry             object
Genre                     object
Director                  object
Box office collection     object
User reviews               int64
Awards                    object
Description               object
Streaming platform        object
dtype: object

In [24]:
df.columns

Index(['Movie name', 'Year of release', 'Watch  hour ', 'Rating', 'Ratedby',
       'Film Industry', 'Genre', 'Director', 'Box office collection',
       'User reviews', 'Awards', 'Description', 'Streaming platform'],
      dtype='object')

In [26]:
col = df.columns
col

Index(['Movie name', 'Year of release', 'Watch  hour ', 'Rating', 'Ratedby',
       'Film Industry', 'Genre', 'Director', 'Box office collection',
       'User reviews', 'Awards', 'Description', 'Streaming platform'],
      dtype='object')

In [33]:
df.columns = [(((i.lower()).strip()).replace("  "," ")).replace(" ","_") for i in list(col)]
df.columns

Index(['movie_name', 'year_of_release', 'watch_hour', 'rating', 'ratedby',
       'film_industry', 'genre', 'director', 'box_office_collection',
       'user_reviews', 'awards', 'description', 'streaming_platform'],
      dtype='object')

In [34]:
df["watch_hour"]

0      2 hours 27 minutes
1                2 hours 
2      2 hours 30 minutes
3      2 hours 25 minutes
4       1 hour 45 minutes
              ...        
245    2 hours 59 minutes
246              2 hours 
247    2 hours 40 minutes
248     2 hours 8 minutes
249    2 hours 33 minutes
Name: watch_hour, Length: 250, dtype: object

In [36]:
df["watch_hour"].unique()

array(['2 hours 27 minutes', '2 hours ', '2 hours 30 minutes',
       '2 hours 25 minutes', '1 hour 45 minutes', '2 hours 40 minutes',
       '2 hours 34 minutes', '2 hours 50 minutes', '2 hours 38 minutes',
       '2 hours 23 minutes', '2 hours 49 minutes', '2 hours 37 minutes',
       '2 hours 15 minutes', '2 hours 44 minutes', '2 hours 4 minutes',
       '2 hours 32 minutes', '2 hours 42 minutes', '2 hours 18 minutes',
       '2 hours 2 minutes', '2 hours 41 minutes', '2 hours 33 minutes',
       '2 hours 46 minutes', '1 hour 57 minutes', '3 hours 12 minutes',
       '2 hours 21 minutes', '2 hours 43 minutes', '3 hours 1minute',
       '2 hours 12 minutes', '1 hour 50 minutes', '2 hours 53 minutes',
       '3 hours 4 minutes', '2 hours 5 minutes', '2 hours 51 minutes',
       '2 hours 26 minutes', '2 hours 36 minutes', '2 hours 28 minutes',
       '3 hours 7 minutes', '5 hours 21 minutes', '3 hours 6 minutes',
       '3 hours 3 minutes', '2 hours 57 minutes', '2 hours 39 minutes',
 

In [42]:
df["watch_hour"].str.split(" ").str[0]

0      2
1      2
2      2
3      2
4      1
      ..
245    2
246    2
247    2
248    2
249    2
Name: watch_hour, Length: 250, dtype: object

In [49]:
df["watch_hour"].str.split(" ").str[0].unique()

array(['2', '1', '3', '5'], dtype=object)

In [43]:
df["watch_hour"].str.split(" ").str[1]

0      hours
1      hours
2      hours
3      hours
4       hour
       ...  
245    hours
246    hours
247    hours
248    hours
249    hours
Name: watch_hour, Length: 250, dtype: object

In [44]:
df["watch_hour"].str.split(" ").str[2]

0      27
1        
2      30
3      25
4      45
       ..
245    59
246      
247    40
248     8
249    33
Name: watch_hour, Length: 250, dtype: object

In [51]:
df["watch_hour"].str.split(" ").str[2].unique()

array(['27', '', '30', '25', '45', '40', '34', '50', '38', '23', '49',
       '37', '15', '44', '4', '32', '42', '18', '2', '41', '33', '46',
       '57', '12', '21', '43', '1minute', '53', '5', '51', '26', '36',
       '28', '7', '6', '3', '39', '35', '20', '56', '19', '9', '16', '54',
       '59', '10', '48', '55', '11', '31', '14', '58', '24', '47', '17',
       '22', '52', '8'], dtype=object)

In [54]:
((df["watch_hour"].str.split(" ").str[2]).replace("",0).replace("1minute",1)).astype(int)

0      27
1       0
2      30
3      25
4      45
       ..
245    59
246     0
247    40
248     8
249    33
Name: watch_hour, Length: 250, dtype: int64

In [55]:
df["total_minutes"] = ((df["watch_hour"].str.split(" ").str[0]).astype(int))*60 + ((df["watch_hour"].str.split(" ").str[2]).replace("",0).replace("1minute",1)).astype(int)
df.head(3)

movie_name  year_of_release          watch_hour  rating ratedby  \
0  12th Fail             2023  2 hours 27 minutes     8.9    126K   
1   Gol Maal             1979            2 hours      8.5     20K   
2   Maharaja             2024  2 hours 30 minutes     8.6     37K   

       film_industry             genre              director  \
0  Bollywood (Hindi)  Drama, Biography    Vidhu Vinod Chopra   
1  Bollywood (Hindi)            Comedy  Hrishikesh Mukherjee   
2  Kollywood (Tamil)      Crime, Drama   Nithilan Saminathan   

  box_office_collection  user_reviews                    awards  \
0           $138,288.00           945  23 wins & 32 nominations   
1                   NIL            48     3 wins & 1 nomination   
2           $975,543.00           370             2 nominations   

                                         description  \
0  The real-life story of IPS Officer Manoj Kumar...   
1  A man's simple lie to secure his job escalates...   
2  A barber seeks vengeance after his home is bur...   

                  streaming_platform  total_minutes  
0                            SonyLIV            147  
1  Amazon Prime Video, YouTube, Zee5            120  
2                            Netflix            150

In [56]:
df["watch_hour"] = df["total_minutes"]/60
df.head(2)

movie_name  year_of_release  watch_hour  rating ratedby      film_industry  \
0  12th Fail             2023        2.45     8.9    126K  Bollywood (Hindi)   
1   Gol Maal             1979        2.00     8.5     20K  Bollywood (Hindi)   

              genre              director box_office_collection  user_reviews  \
0  Drama, Biography    Vidhu Vinod Chopra           $138,288.00           945   
1            Comedy  Hrishikesh Mukherjee                   NIL            48   

                     awards  \
0  23 wins & 32 nominations   
1     3 wins & 1 nomination   

                                         description  \
0  The real-life story of IPS Officer Manoj Kumar...   
1  A man's simple lie to secure his job escalates...   

                  streaming_platform  total_minutes  
0                            SonyLIV            147  
1  Amazon Prime Video, YouTube, Zee5            120

In [57]:
df.dtypes

movie_name                object
year_of_release            int64
watch_hour               float64
rating                   float64
ratedby                   object
film_industry             object
genre                     object
director                  object
box_office_collection     object
user_reviews               int64
awards                    object
description               object
streaming_platform        object
total_minutes              int64
dtype: object

In [59]:
df["ratedby"].unique()

array(['126K', '20K', '37K', '25K', '17K', '26K', '19K', '441K', '16K',
       '22K', '13K', '59K', '41K', '8.8K', '8.2K', '209K', '5.4K', '39K',
       '215K', '23K', '125K', '38K', '7.3K', '12K', '42K', '44K', '5.9K',
       '34K', '219K', '71K', '45K', '15K', '5.7K', '24K', '6.5K', '97K',
       '105K', '52K', '73K', '8.7K', '48K', '6.2K', '87K', '7.4K', '18K',
       '104K', '8.6K', '9.3K', '89K', '99K', '7K', '65K', '61K', '51K',
       '86K', '36K', '14K', '5.5K', '204K', '8.1K', '122K', '76K', '27K',
       '46K', '7.8K', '88K', '8K', '66K', '11K', '253K', '69K', '32K',
       '7.6K', '123K', '83K', '117K', '7.5K', '49K', '10K', '6.4K', '9K',
       '8.9K', '78K', '60K', '7.7K', '131K', '58K', '8.5K', '28K', '56K',
       '100K', '29K', '9.1K', '6.1K', '9.6K', '7.9K', '6.6K', '6.3K',
       '8.4K', '6.8K', '5.3K', '152K', '9.9K', '33K', '40K', '139K',
       '62K', '8.3K', '6.9K', '9.7K'], dtype=object)

In [65]:
df["ratedby"].str.split("K")

0      [126, ]
1       [20, ]
2       [37, ]
3       [25, ]
4       [17, ]
        ...   
245     [17, ]
246    [5.3, ]
247     [17, ]
248    [6.2, ]
249    [6.3, ]
Name: ratedby, Length: 250, dtype: object

In [66]:
type(df["ratedby"].str.split("K"))

pandas.core.series.Series

In [70]:
df["ratedby_in_k"] = (df["ratedby"].str.split("K").str[0]).astype(float)

In [71]:
df.head(2)

movie_name  year_of_release  watch_hour  rating ratedby      film_industry  \
0  12th Fail             2023        2.45     8.9    126K  Bollywood (Hindi)   
1   Gol Maal             1979        2.00     8.5     20K  Bollywood (Hindi)   

              genre              director box_office_collection  user_reviews  \
0  Drama, Biography    Vidhu Vinod Chopra           $138,288.00           945   
1            Comedy  Hrishikesh Mukherjee                   NIL            48   

                     awards  \
0  23 wins & 32 nominations   
1     3 wins & 1 nomination   

                                         description  \
0  The real-life story of IPS Officer Manoj Kumar...   
1  A man's simple lie to secure his job escalates...   

                  streaming_platform  total_minutes  ratedby_in_k  
0                            SonyLIV            147         126.0  
1  Amazon Prime Video, YouTube, Zee5            120          20.0

In [72]:
df["ratedby"] = (df["ratedby_in_k"]*1000).astype(int)
df.dtypes

movie_name                object
year_of_release            int64
watch_hour               float64
rating                   float64
ratedby                    int64
film_industry             object
genre                     object
director                  object
box_office_collection     object
user_reviews               int64
awards                    object
description               object
streaming_platform        object
total_minutes              int64
ratedby_in_k             float64
dtype: object

In [73]:
df.drop(["total_minutes","ratedby_in_k"],axis=1,inplace=True)

In [74]:
df.head(2)

movie_name  year_of_release  watch_hour  rating  ratedby      film_industry  \
0  12th Fail             2023        2.45     8.9   126000  Bollywood (Hindi)   
1   Gol Maal             1979        2.00     8.5    20000  Bollywood (Hindi)   

              genre              director box_office_collection  user_reviews  \
0  Drama, Biography    Vidhu Vinod Chopra           $138,288.00           945   
1            Comedy  Hrishikesh Mukherjee                   NIL            48   

                     awards  \
0  23 wins & 32 nominations   
1     3 wins & 1 nomination   

                                         description  \
0  The real-life story of IPS Officer Manoj Kumar...   
1  A man's simple lie to secure his job escalates...   

                  streaming_platform  
0                            SonyLIV  
1  Amazon Prime Video, YouTube, Zee5

In [77]:
df["box_office_collection"].unique()

array(['$138,288.00', ' NIL', '$975,543.00', '$120,481.93', '$134,241.00',
       '$60,262,836.00', '$1,610,897.00', '$843,373.49', '$398,615.00',
       '$1,312,005.00', '$7,523,995.00', '$181,001.00', '$21,897,373.00',
       '$108,886.00', '$303,723,636.00', '$204,669.00', '$1,807,228.92',
       '$616,333.00', '$633.00', '$1,943.00', '$3,841.00', '$522,938.00',
       '$379,728.00', '$2,208,029.00', '$13,076,265.00', '$3,373,493.98',
       '$40,194.00', '$472,865.00', '$2,409,638.55', '$135,342.00',
       '$180,217.00', '$11,939.76', '$7,228,915.66', '$4,454.00',
       '$9,400,000.00', '$4,384,642.00', '$17,551,296.00', '$279,078.00',
       '$3,417.00', '$5,192,743.00', '$273,191.00', '$62,475,342.00',
       '$6,024,096.39', '$377,803.00', '$3,137,716.00', '$1,953,889.00',
       '$62,313.00', '$3,311,391.00', '$1,691,980.00', '$586,973.00',
       '$546,792.00', '$3,784,106.00', '$40,311.00', '$602,409.64',
       '$21,505,244.00', '$1,266,341.00', '$400,000.00',
       '$100

In [81]:
df[df["box_office_collection"] == " NIL"].index

Index([  1,   5,   6,   8,  15,  17,  27,  28,  29,  31,  32,  36,  39,  40,
        44,  49,  52,  54,  55,  58,  60,  65,  66,  67,  69,  78,  92,  95,
       101, 102, 116, 119, 120, 125, 127, 130, 132, 140, 141, 147, 151, 154,
       162, 173, 191, 198, 211, 221, 232, 237, 245, 246],
      dtype='int64')

### This will work as missing value and we will drop this rows which contaning the value as NIL.

In [83]:
df.drop(df[df["box_office_collection"] == " NIL"].index, axis=0,inplace= True)

In [84]:
df["box_office_collection"].unique()

array(['$138,288.00', '$975,543.00', '$120,481.93', '$134,241.00',
       '$60,262,836.00', '$1,610,897.00', '$843,373.49', '$398,615.00',
       '$1,312,005.00', '$7,523,995.00', '$181,001.00', '$21,897,373.00',
       '$108,886.00', '$303,723,636.00', '$204,669.00', '$1,807,228.92',
       '$616,333.00', '$633.00', '$1,943.00', '$3,841.00', '$522,938.00',
       '$379,728.00', '$2,208,029.00', '$13,076,265.00', '$3,373,493.98',
       '$40,194.00', '$472,865.00', '$2,409,638.55', '$135,342.00',
       '$180,217.00', '$11,939.76', '$7,228,915.66', '$4,454.00',
       '$9,400,000.00', '$4,384,642.00', '$17,551,296.00', '$279,078.00',
       '$3,417.00', '$5,192,743.00', '$273,191.00', '$62,475,342.00',
       '$6,024,096.39', '$377,803.00', '$3,137,716.00', '$1,953,889.00',
       '$62,313.00', '$3,311,391.00', '$1,691,980.00', '$586,973.00',
       '$546,792.00', '$3,784,106.00', '$40,311.00', '$602,409.64',
       '$21,505,244.00', '$1,266,341.00', '$400,000.00',
       '$100,370,104

In [86]:
df["box_office_collection"] = df["box_office_collection"].str.replace("$","")

In [88]:
df["box_office_collection"] = df["box_office_collection"].str.replace(",","")
df["box_office_collection"] = df["box_office_collection"].str.replace(".","")
df["box_office_collection"].unique()

array(['13828800', '97554300', '12048193', '13424100', '6026283600',
       '161089700', '84337349', '39861500', '131200500', '752399500',
       '18100100', '2189737300', '10888600', '30372363600', '20466900',
       '180722892', '61633300', '63300', '194300', '384100', '52293800',
       '37972800', '220802900', '1307626500', '337349398', '4019400',
       '47286500', '240963855', '13534200', '18021700', '1193976',
       '722891566', '445400', '940000000', '438464200', '1755129600',
       '27907800', '341700', '519274300', '27319100', '6247534200',
       '602409639', '37780300', '313771600', '195388900', '6231300',
       '331139100', '169198000', '58697300', '54679200', '378410600',
       '4031100', '60240964', '2150524400', '126634100', '40000000',
       '10037010400', '465200', '172480600', '313515000', '2983763700',
       '28873300', '2427400', '687909800', '297334700', '24655100',
       '449013400', '3934657700', '574638554', '43486200', '73451400',
       '1214008700', '

In [89]:
df["box_office_collection_in_doller"] = df["box_office_collection"].astype("int")
df.drop("box_office_collection",axis=1,inplace= True)
df.head(2)

movie_name  year_of_release  watch_hour  rating  ratedby      film_industry  \
0  12th Fail             2023        2.45     8.9   126000  Bollywood (Hindi)   
2   Maharaja             2024        2.50     8.6    37000  Kollywood (Tamil)   

              genre             director  user_reviews  \
0  Drama, Biography   Vidhu Vinod Chopra           945   
2      Crime, Drama  Nithilan Saminathan           370   

                     awards  \
0  23 wins & 32 nominations   
2             2 nominations   

                                         description streaming_platform  \
0  The real-life story of IPS Officer Manoj Kumar...            SonyLIV   
2  A barber seeks vengeance after his home is bur...            Netflix   

   box_office_collection_in_doller  
0                         13828800  
2                         97554300

In [91]:
df["box_office_collection_in_doller"] = df["box_office_collection_in_doller"]/100
df.head(2)

movie_name  year_of_release  watch_hour  rating  ratedby      film_industry  \
0  12th Fail             2023        2.45     8.9   126000  Bollywood (Hindi)   
2   Maharaja             2024        2.50     8.6    37000  Kollywood (Tamil)   

              genre             director  user_reviews  \
0  Drama, Biography   Vidhu Vinod Chopra           945   
2      Crime, Drama  Nithilan Saminathan           370   

                     awards  \
0  23 wins & 32 nominations   
2             2 nominations   

                                         description streaming_platform  \
0  The real-life story of IPS Officer Manoj Kumar...            SonyLIV   
2  A barber seeks vengeance after his home is bur...            Netflix   

   box_office_collection_in_doller  
0                         138288.0  
2                         975543.0

In [93]:
df.dtypes

movie_name                          object
year_of_release                      int64
watch_hour                         float64
rating                             float64
ratedby                              int64
film_industry                       object
genre                               object
director                            object
user_reviews                         int64
awards                              object
description                         object
streaming_platform                  object
box_office_collection_in_doller    float64
dtype: object

In [95]:
df["awards"].unique()

array(['23 wins & 32 nominations', '2 nominations',
       '7 wins & 1 nomination', '4 wins & 2 nominations total',
       '64 wins & 30 nominations', '1 win & 7 nominations', '5 wins',
       '6 wins & 19 nominations', '13 wins & 11 nominations',
       '3 wins & 6 nominations', '1 win', '28 wins & 18 nominations',
       '5 nominations', '41 wins & 38 nominations',
       '9 wins & 7 nominations', '23 wins & 24 nominations',
       '14 wins & 11 nominations', '2 wins & 5 nominations',
       '11 wins & 7 nominations', '10 nominations',
       '15 wins & 8 nominations', '22 wins & 9 nominations',
       '11 wins & 24 nominations', '16 wins & 6 nominations', '18 wins',
       '2 wins & 7 nominations', '6 wins & 2 nominations total',
       '9 wins & 8 nominations', '17 wins & 10 nominations',
       '11 wins & 2 nominations total', '20 wins & 2 nominations', 'NIL',
       '25 wins & 5 nominations', '1 win & 4 nominations',
       '2 wins & 8 nominations', '18 wins & 65 nominations',
  

In [99]:
'6 wins'.split("&")

['6 wins']

In [101]:
import re

In [103]:
df["awards"][:10]

0         23 wins & 32 nominations
2                    2 nominations
3            7 wins & 1 nomination
4     4 wins & 2 nominations total
7         64 wins & 30 nominations
9            1 win & 7 nominations
10                          5 wins
11         6 wins & 19 nominations
12        13 wins & 11 nominations
13          3 wins & 6 nominations
Name: awards, dtype: object

In [115]:
df['wins'] = [int(re.search(r'(\d+)\s+wins', x).group(1)) if re.search(r'(\d+)\s+wins', x) else 0 for x in df['awards']]
df['nominations'] = [int(re.search(r'(\d+)\s+nominations', x).group(1)) if re.search(r'(\d+)\s+nominations', x) else 0 for x in df['awards']]


In [116]:
df.head(2)

movie_name  year_of_release  watch_hour  rating  ratedby      film_industry  \
0  12th Fail             2023        2.45     8.9   126000  Bollywood (Hindi)   
2   Maharaja             2024        2.50     8.6    37000  Kollywood (Tamil)   

              genre             director  user_reviews  \
0  Drama, Biography   Vidhu Vinod Chopra           945   
2      Crime, Drama  Nithilan Saminathan           370   

                     awards  \
0  23 wins & 32 nominations   
2             2 nominations   

                                         description streaming_platform  \
0  The real-life story of IPS Officer Manoj Kumar...            SonyLIV   
2  A barber seeks vengeance after his home is bur...            Netflix   

   box_office_collection_in_doller  wins  nominations  
0                         138288.0    23           32  
2                         975543.0     0            2

In [118]:
df["total_awards"] = df["wins"] + df["nominations"]

In [120]:
df.drop("awards",axis=1,inplace=True)

In [121]:
df.dtypes

movie_name                          object
year_of_release                      int64
watch_hour                         float64
rating                             float64
ratedby                              int64
film_industry                       object
genre                               object
director                            object
user_reviews                         int64
description                         object
streaming_platform                  object
box_office_collection_in_doller    float64
wins                                 int64
nominations                          int64
total_awards                         int64
dtype: object

### Now all the columns are in their relevant datatype.

In [122]:
df.shape

(198, 15)

In [123]:
df.isnull().sum()

movie_name                         0
year_of_release                    0
watch_hour                         0
rating                             0
ratedby                            0
film_industry                      0
genre                              0
director                           0
user_reviews                       0
description                        0
streaming_platform                 0
box_office_collection_in_doller    0
wins                               0
nominations                        0
total_awards                       0
dtype: int64

### We drop the rows that does not contain the box_office_collection feature value and we handle the missing value of the feature awards by filling it with 0.

In [125]:
df[df.duplicated()]

Empty DataFrame
Columns: [movie_name, year_of_release, watch_hour, rating, ratedby, film_industry, genre, director, user_reviews, description, streaming_platform, box_office_collection_in_doller, wins, nominations, total_awards]
Index: []

## Dataset is not containing the duplicated values.

### **Q 3. Filtering and Sorting**
  > a. Filter the dataset to display only movies released after the year 2000.

  > b. Sort the DataFrame by IMDB Rating in descending order.

  > c. Create a new DataFrame that contains only movies with an IMDB rating greater than 8.5.


In [126]:
df.head(3)

movie_name  year_of_release  watch_hour  rating  ratedby      film_industry  \
0  12th Fail             2023    2.450000     8.9   126000  Bollywood (Hindi)   
2   Maharaja             2024    2.500000     8.6    37000  Kollywood (Tamil)   
3    Nayakan             1987    2.416667     8.7    25000  Kollywood (Tamil)   

              genre             director  user_reviews  \
0  Drama, Biography   Vidhu Vinod Chopra           945   
2      Crime, Drama  Nithilan Saminathan           370   
3      Crime, Drama          Mani Ratnam           237   

                                         description  \
0  The real-life story of IPS Officer Manoj Kumar...   
2  A barber seeks vengeance after his home is bur...   
3  A common man's struggles against a corrupt pol...   

            streaming_platform  box_office_collection_in_doller  wins  \
0                      SonyLIV                        138288.00    23   
2                      Netflix                        975543.00     0   
3  Amazon Prime Video, YouTube                        120481.93     7   

   nominations  total_awards  
0           32            55  
2            2             2  
3            0             7

In [128]:
df_after_2000 = df[df["year_of_release"] > 2000]
df_after_2000

movie_name  year_of_release  watch_hour  rating  ratedby  \
0                     12th Fail             2023    2.450000     8.9   126000   
2                      Maharaja             2024    2.500000     8.6    37000   
7                      3 Idiots             2009    2.833333     8.4   441000   
9                  Black Friday             2004    2.383333     8.4    22000   
11   Rocketry: The Nambi Effect             2022    2.616667     8.7    59000   
..                          ...              ...         ...     ...      ...   
243                 Vicky Donor             2012    2.100000     7.7    46000   
244            Angamaly Diaries             2017    2.200000     7.9     6800   
247                      Ghilli             2004    2.666667     8.1    17000   
248    Mukundan Unni Associates             2022    2.133333     7.9     6200   
249                  Happy Days             2007    2.550000     7.9     6300   

             film_industry                   genre              director  \
0        Bollywood (Hindi)        Drama, Biography    Vidhu Vinod Chopra   
2        Kollywood (Tamil)            Crime, Drama   Nithilan Saminathan   
7        Bollywood (Hindi)           Comedy, Drama       Rajkumar Hirani   
9        Bollywood (Hindi)  Crime, Drama, Thriller        Anurag Kashyap   
11       Bollywood (Hindi)        Biography, Drama              Madhavan   
..                     ...                     ...                   ...   
243      Bollywood (Hindi)           Comedy, Drama        Shoojit Sircar   
244  Mollywood (Malayalam)            Crime, Drama  Lijo Jose Pellissery   
247      Kollywood (Tamil)           Action, Drama               Dharani   
248  Mollywood (Malayalam)           Satire, Drama  Abhinav Sunder Nayak   
249     Tollywood (Telugu)  Drama, Romance, Comedy        Sekhar Kammula   

     user_reviews                                        description  \
0             945  The real-life story of IPS Officer Manoj Kumar...   
2             370  A barber seeks vengeance after his home is bur...   
7            1000  Two friends are searching for their long lost ...   
9              88  A film about the investigations following the ...   
11           1100  The story of Indian Space Research Organizatio...   
..            ...                                                ...   
243            82  A man is brought in by an infertility doctor t...   
244            39  Vincent Pepe who wanted to be a powerful leade...   
247            47  Velu, an aspiring kabaddi player, goes to Madu...   
248            62  Advocate Mukundan Unni, played by Vineeth Sree...   
249            16  A group of eight youngsters from an engineerin...   

              streaming_platform  box_office_collection_in_doller  wins  \
0                        SonyLIV                        138288.00    23   
2                        Netflix                        975543.00     0   
7    Netflix, Amazon Prime Video                      60262836.00    64   
9                        YouTube                       1610897.00     0   
11            Amazon Prime Video                        398615.00     6   
..                           ...                              ...   ...   
243  Amazon Prime Video, Netflix                       6456358.00    43   
244                      Netflix                        661915.00     2   
247              Disney+ Hotstar                        163961.00     2   
248              Disney+ Hotstar                         44990.00     0   
249           Amazon Prime Video                        361445.78     7   

     nominations  total_awards  
0             32            55  
2              2             2  
7             30            94  
9              7             7  
11            19            25  
..           ...           ...  
243           37            80  
244            6             8  
247            2             4  
248            0             0  
249   

In [129]:
df_after_2000.shape

(174, 15)

In [141]:
df_sort_rating = df.sort_values("rating",ascending=False)
df_sort_rating.head(15)

movie_name  year_of_release  watch_hour  rating  ratedby  \
0                     12th Fail             2023    2.450000     8.9   126000   
14                     Kireedam             1989    2.066667     8.9     8800   
23                    Natsamrat             2016    2.766667     8.8     7300   
11   Rocketry: The Nambi Effect             2022    2.616667     8.7    59000   
21              Soorarai Pottru             2020    2.550000     8.7   125000   
57                      Peranbu             2018    2.450000     8.7    17000   
13                  777 Charlie             2022    2.733333     8.7    41000   
10            Manichithrathazhu             1993    2.816667     8.7    13000   
131   A Northern Story of Valor             1989    2.800000     8.7     6400   
3                       Nayakan             1987    2.416667     8.7    25000   
71                     Spadikam             1995    2.500000     8.6     7000   
2                      Maharaja             2024    2.500000     8.6    37000   
12            Kumbalangi Nights             2019    2.250000     8.5    19000   
83              Aaranya Kaandam             2010    2.100000     8.5     5500   
33                   Sita Ramam             2022    2.716667     8.5    71000   

             film_industry                    genre                director  \
0        Bollywood (Hindi)         Drama, Biography      Vidhu Vinod Chopra   
14   Mollywood (Malayalam)            Drama, Family            Sibi Malayil   
23          Marathi Cinema           Drama, Tragedy        Mahesh Manjrekar   
11       Bollywood (Hindi)         Biography, Drama                Madhavan   
21       Kollywood (Tamil)         Drama, Biography           Sudha Kongara   
57       Kollywood (Tamil)                    Drama                     Ram   
13    Sandalwood (Kannada)         Adventure, Drama              Kiranraj K   
10   Mollywood (Malayalam)  Horror, Thriller, Drama                   Fazil   
131      Kollywood (Tamil)            Drama, Action            T. Hariharan   
3        Kollywood (Tamil)             Crime, Drama             Mani Ratnam   
71   Mollywood (Malayalam)            Action, Drama                 Bhadran   
2        Kollywood (Tamil)             Crime, Drama     Nithilan Saminathan   
12   Mollywood (Malayalam)            Drama, Family      Madhu C. Narayanan   
83       Kollywood (Tamil)       Neo-noir, Thriller  Thiagarajan Kumararaja   
33      Tollywood (Telugu)           Romance, Drama        Hanu Raghavapudi   

     user_reviews                                        description  \
0             945  The real-life story of IPS Officer Manoj Kumar...   
14             33  The life of a young man turns upside down when...   
23             72  After he retires, an aging theatre actor and h...   
11           1100  The story of Indian Space Research Organizatio...   
21           2100  Nedumaaran Rajangam "Maara" sets out to make t...   
57            425  A single father tries to raise his daughter, w...   
13            392  Dharma is stuck in a rut with his negative and...   
10             53  When a forbidden room in an old bungalow is un...   
131            45  The film offers an alternative version of the ...   
3             237  A common man's struggles against a corrupt pol...   
71             19  A small-town criminal constantly clashes with ...   
2             370  A barber seeks vengeance after his home is bur...   
12            194  The film revolves around four brothers who sha...   
83             25  A mob boss must deal with a disgruntled mistre...   
33            701  An orphan soldier, Lieutenant Ram's life chang...   

              streaming_platform  box_office_collection_in_doller  wins  \
0                        SonyLIV                        138288.00    23   
14            Amazon Prime Video                        181001.00     0   
23                          Zee5                           633.00     2   
11    

In [135]:
movies_8_5 = df[df["rating"] > 8.5]["movie_name"]
movies_8_5

0                       12th Fail
2                        Maharaja
3                         Nayakan
10              Manichithrathazhu
11     Rocketry: The Nambi Effect
13                    777 Charlie
14                       Kireedam
21                Soorarai Pottru
23                      Natsamrat
57                        Peranbu
71                       Spadikam
131     A Northern Story of Valor
Name: movie_name, dtype: object

In [142]:
movies_8_5.shape

(12,)

In [139]:
df_movies = pd.DataFrame(list(movies_8_5),columns=["Movies_name"])
df_movies

Movies_name
0                    12th Fail
1                     Maharaja
2                      Nayakan
3            Manichithrathazhu
4   Rocketry: The Nambi Effect
5                  777 Charlie
6                     Kireedam
7              Soorarai Pottru
8                    Natsamrat
9                      Peranbu
10                    Spadikam
11   A Northern Story of Valor

In [140]:
df_movies.shape

(12, 1)

In [143]:
type(df_movies)

pandas.core.frame.DataFrame

### **Q 4. Group By Operations**
  > a. Group the movies by the Year column and calculate the average IMDB rating for each year.

  > b. Group the data by Director and count the number of movies each director has in the top 250.

  > c. Find the highest-rated movie for each year by grouping the data by Year and selecting the movie with the highest rating in each group.


In [145]:
df.to_csv("clean_data.csv",index=False)
d = pd.read_csv("clean_data.csv")
d.head(2)

movie_name  year_of_release  watch_hour  rating  ratedby      film_industry  \
0  12th Fail             2023        2.45     8.9   126000  Bollywood (Hindi)   
1   Maharaja             2024        2.50     8.6    37000  Kollywood (Tamil)   

              genre             director  user_reviews  \
0  Drama, Biography   Vidhu Vinod Chopra           945   
1      Crime, Drama  Nithilan Saminathan           370   

                                         description streaming_platform  \
0  The real-life story of IPS Officer Manoj Kumar...            SonyLIV   
1  A barber seeks vengeance after his home is bur...            Netflix   

   box_office_collection_in_doller  wins  nominations  total_awards  
0                         138288.0    23           32            55  
1                         975543.0     0            2             2

In [144]:
df.head()

movie_name  year_of_release  watch_hour  rating  ratedby  \
0         12th Fail             2023    2.450000     8.9   126000   
2          Maharaja             2024    2.500000     8.6    37000   
3           Nayakan             1987    2.416667     8.7    25000   
4  The World of Apu             1959    1.750000     8.4    17000   
7          3 Idiots             2009    2.833333     8.4   441000   

       film_industry             genre             director  user_reviews  \
0  Bollywood (Hindi)  Drama, Biography   Vidhu Vinod Chopra           945   
2  Kollywood (Tamil)      Crime, Drama  Nithilan Saminathan           370   
3  Kollywood (Tamil)      Crime, Drama          Mani Ratnam           237   
4     Bengali Cinema             Drama         Satyajit Ray            62   
7  Bollywood (Hindi)     Comedy, Drama      Rajkumar Hirani          1000   

                                         description  \
0  The real-life story of IPS Officer Manoj Kumar...   
2  A barber seeks vengeance after his home is bur...   
3  A common man's struggles against a corrupt pol...   
4  This final installment in Satyajit Ray's Apu T...   
7  Two friends are searching for their long lost ...   

            streaming_platform  box_office_collection_in_doller  wins  \
0                      SonyLIV                        138288.00    23   
2                      Netflix                        975543.00     0   
3  Amazon Prime Video, YouTube                        120481.93     7   
4  Amazon Prime Video, Hoichoi                        134241.00     4   
7  Netflix, Amazon Prime Video                      60262836.00    64   

   nominations  total_awards  
0           32            55  
2            2             2  
3            0             7  
4            2             6  
7           30            94

In [148]:
avg_per_year = df.groupby("year_of_release").agg({"rating":"mean"})
avg_per_year

rating
year_of_release          
1955             8.200000
1956             8.200000
1958             7.900000
1959             8.400000
1960             8.100000
1971             8.100000
1975             8.150000
1987             8.700000
1988             7.900000
1989             8.800000
1993             8.700000
1995             8.300000
1996             8.100000
1997             7.900000
1999             8.175000
2000             8.100000
2001             8.100000
2002             8.000000
2003             8.016667
2004             8.071429
2005             8.200000
2006             8.062500
2007             8.020000
2008             8.150000
2009             8.075000
2010             8.150000
2011             8.075000
2012             8.061538
2013             8.080000
2014             8.185714
2015             8.158333
2016             8.200000
2017             8.050000
2018             8.260000
2019             8.193750
2020             8.200000
2021             8.200000
2022             8.253846
2023             8.250000
2024             8.225000

In [152]:
director_movies_count = df.groupby("director").agg({"movie_name":"count"})
director_movies_count.columns = ["count"]
director_movies_count.head(2)

count
director              
A.L. Vijay           1
A.R. Murugadoss      1

In [155]:
director_movies_count = director_movies_count.sort_values("count",ascending=False)
director_movies_count.head(10)

count
director                    
Anurag Kashyap             6
Vetrimaaran                5
Jeethu Joseph              4
Satyajit Ray               4
Gautham Vasudev Menon      4
Rajkumar Hirani            4
Neeraj Pandey              3
Lokesh Kanagaraj           3
S. Shankar                 3
Karthik Subbaraj           3

In [173]:
df.groupby("year_of_release")["rating"]

In [172]:
df.groupby("year_of_release")["rating"].idxmax()

year_of_release
1955     43
1956     38
1958    196
1959      4
1960    138
1971    128
1975     46
1987      3
1988    192
1989     14
1993     10
1995     71
1996    134
1997    182
1999     63
2000    124
2001     86
2002    159
2003    115
2004      9
2005     35
2006     85
2007     16
2008    144
2009      7
2010     83
2011     59
2012     51
2013     34
2014    174
2015     24
2016     23
2017     75
2018     57
2019     12
2020     21
2021     25
2022     11
2023      0
2024      2
Name: rating, dtype: int64

In [176]:
df.loc[df.groupby("year_of_release")["rating"].idxmax()][["year_of_release","rating"]].reset_index(drop=True)

year_of_release  rating
0              1955     8.2
1              1956     8.2
2              1958     7.9
3              1959     8.4
4              1960     8.1
5              1971     8.1
6              1975     8.3
7              1987     8.7
8              1988     7.9
9              1989     8.9
10             1993     8.7
11             1995     8.6
12             1996     8.1
13             1997     7.9
14             1999     8.4
15             2000     8.3
16             2001     8.1
17             2002     8.0
18             2003     8.3
19             2004     8.4
20             2005     8.3
21             2006     8.4
22             2007     8.3
23             2008     8.2
24             2009     8.4
25             2010     8.5
26             2011     8.2
27             2012     8.2
28             2013     8.3
29             2014     8.4
30             2015     8.4
31             2016     8.8
32             2017     8.2
33             2018     8.7
34             2019     8.5
35             2020     8.7
36             2021     8.4
37             2022     8.7
38             2023     8.9
39             2024     8.6

### **Q 1. Column Creation and Manipulation**
  > a. Create a new column named Rating Category that categorizes movies as "Excellent" if the IMDB rating is 9.0 or above, "Good" if between 8.0 and 9.0, and "Average" if below 8.0.

  > b. Extract the Year from the movie title (assuming the year is part of the title) and create a new column Extracted Year.

  > c. Create a new column that combines the Title and Director into a single string, separated by a hyphen.

In [178]:
def categorize_value(val):
  if val >= 9:
    return "Excellent"
  elif val >=8 :
    return "Good"
  else :
    return "Average"

In [180]:
df["rating_category"] = df["rating"].map(categorize_value)
df.head(2)

movie_name  year_of_release  watch_hour  rating  ratedby      film_industry  \
0  12th Fail             2023        2.45     8.9   126000  Bollywood (Hindi)   
2   Maharaja             2024        2.50     8.6    37000  Kollywood (Tamil)   

              genre             director  user_reviews  \
0  Drama, Biography   Vidhu Vinod Chopra           945   
2      Crime, Drama  Nithilan Saminathan           370   

                                         description streaming_platform  \
0  The real-life story of IPS Officer Manoj Kumar...            SonyLIV   
2  A barber seeks vengeance after his home is bur...            Netflix   

   box_office_collection_in_doller  wins  nominations  total_awards  \
0                         138288.0    23           32            55   
2                         975543.0     0            2             2   

  rating_category  
0            Good  
2            Good

In [182]:
df["movie_name"].unique()

array(['12th Fail', 'Maharaja', 'Nayakan', 'The World of Apu', '3 Idiots',
       'Black Friday', 'Manichithrathazhu', 'Rocketry: The Nambi Effect',
       'Kumbalangi Nights', '777 Charlie', 'Kireedam',
       'Like Stars on Earth', 'Laapataa Ladies', 'Dangal', 'Jersey',
       'Soorarai Pottru', '96', 'Natsamrat', 'Visaaranai', 'Drishyam 2',
       'Kaithi', 'Asuran', 'Sita Ramam', 'Drishyam', 'Anniyan',
       'Khosla Ka Ghosla!', 'Aparajito', 'Vada Chennai', 'Thani Oruvan',
       'Pather Panchali', 'Bangalore Days', 'Chupke Chupke', 'Premam',
       'Agent Sai Srinivasa Athreya', 'Gangs of Wasseypur',
       'Bhaag Milkha Bhaag', 'Mahanati', 'Peranbu',
       'Zindagi Na Milegi Dobara', 'Super Deluxe', 'Andhadhun',
       'Mudhalvan', 'Shahid', 'Munna Bhai M.B.B.S.', 'Swades', 'Spadikam',
       'Chhichhore', 'Tumbbad', 'Papanasam', 'Vikram Vedha',
       'Manjummel Boys', 'Paan Singh Tomar', 'Mandela', 'Chak De! India',
       'Black', 'Aruvi', 'Aaranya Kaandam', 'PK', 'Pudhu Pet

In [185]:
df["extracted_year"] = [int(re.search("(\d+)",i).group(1)) if re.search("(\d+)",i) else 0 for i in df["movie_name"] ]
df.head(2)

movie_name  year_of_release  watch_hour  rating  ratedby      film_industry  \
0  12th Fail             2023        2.45     8.9   126000  Bollywood (Hindi)   
2   Maharaja             2024        2.50     8.6    37000  Kollywood (Tamil)   

              genre             director  user_reviews  \
0  Drama, Biography   Vidhu Vinod Chopra           945   
2      Crime, Drama  Nithilan Saminathan           370   

                                         description streaming_platform  \
0  The real-life story of IPS Officer Manoj Kumar...            SonyLIV   
2  A barber seeks vengeance after his home is bur...            Netflix   

   box_office_collection_in_doller  wins  nominations  total_awards  \
0                         138288.0    23           32            55   
2                         975543.0     0            2             2   

  rating_category  extracted_year  
0            Good              12  
2            Good               0

In [187]:
df["movie_director"] = df["movie_name"] + " - " + df["director"]
df.head(2)

movie_name  year_of_release  watch_hour  rating  ratedby      film_industry  \
0  12th Fail             2023        2.45     8.9   126000  Bollywood (Hindi)   
2   Maharaja             2024        2.50     8.6    37000  Kollywood (Tamil)   

              genre             director  user_reviews  \
0  Drama, Biography   Vidhu Vinod Chopra           945   
2      Crime, Drama  Nithilan Saminathan           370   

                                         description streaming_platform  \
0  The real-life story of IPS Officer Manoj Kumar...            SonyLIV   
2  A barber seeks vengeance after his home is bur...            Netflix   

   box_office_collection_in_doller  wins  nominations  total_awards  \
0                         138288.0    23           32            55   
2                         975543.0     0            2             2   

  rating_category  extracted_year                  movie_director  
0            Good              12  12th Fail - Vidhu Vinod Chopra  
2            Good               0  Maharaja - Nithilan Saminathan